### PR TITLE
Optimization: cache semantic lookup in tokenizer, add semantic tokens bench marks

### PR DESCRIPTION
**Pull request by soulomoon**
1. It is high chance that an identifier might appear multiple times in a file, cache the identifier hs semantic type result in tokenizer saves repeated computation.
2. A new test case is added to ensure invisible token won't not sabotage the cache lookup
3. Add bench mark for semantic tokens
4. fix bench mark running issue.

It gains about 5% performance increasement to the GetSemanticTokens Rule.

### CODE REVIEW

**Comment by wz1000** [comment]
could we have some benchmarks on large files to see if this actually makes a difference?

**Comment by soulomoon** [comment]
WIth 
```
  - name: cabal
    package: Cabal
    version: 3.6.3.0
    modules:
        - src/Distribution/Simple/Configure.hs
```

`cabal bench -j --benchmark-options="profiled-cabal"`

| version               | configuration | name           | success | samples | startup | setup | userT | delayedT | 1stBuildT | avgPerRespT | totalT | rulesBuilt | rulesChanged | rulesVisited | rulesTotal | ruleEdges | ghcRebuilds | maxResidency | allocatedBytes |
| --------------------- | ------------- | -------------- | ------- | ------- | ------- | ----- | ----- | -------- | --------- | ----------- | ------ | ---------- | ------------ | ------------ | ---------- | --------- | ----------- | ------------ | -------------- |
| upstream              | All           | semanticTokens | True    | 50      | 2.09    | 0.00  | 9.37  | 0.03     | 3.10      | 0.13        | 9.40   | 5147       | 5146         | 5147         | 5908       | 31251     | 1           | 174MB        | 13491MB        |
| cache-semantic-lookup | All           | semanticTokens | True    | 50      | 1.87    | 0.00  | 8.36  | 0.10     | 2.77      | 0.11        | 8.46   | 5147       | 5146         | 5147         | 5908       | 31251     | 1           | 173MB        | 13437MB        |
|                       |               |                |         |         |         |       |       |          |           |             |        |            |              |              |            |           |             |              |                |

**Comment by wz1000** [comment]
Looks like a decent improvement. Could you add this benchmark to the ghcide-bench experiments please?

**Comment by wz1000** [comment]
the `ghcRebuilds` number seems to indicate that we aren't editing the file in the benchmark passes. Could you ensure that each benchmark iteration also edits the file?

**Comment by soulomoon** [comment]
> the `ghcRebuilds` number seems to indicate that we aren't editing the file in the benchmark passes. Could you ensure that each benchmark iteration also edits the file?

Sure, I will make some changes

**Comment by soulomoon** [comment]
I need to took some more time to come up a more decent benchmark on this. Convert it to draft now.

**Comment by soulomoon** [comment]
The overal improvement is does really obivious. Slightly better mem usage and speed.

| version               | configuration | name           | success | samples | startup | setup | userT  | delayedT | 1stBuildT | avgPerRespT | totalT | rulesBuilt | rulesChanged | rulesVisited | rulesTotal | ruleEdges | ghcRebuilds | maxResidency | allocatedBytes |
| --------------------- | ------------- | -------------- | ------- | ------- | ------- | ----- | ------ | -------- | --------- | ----------- | ------ | ---------- | ------------ | ------------ | ---------- | --------- | ----------- | ------------ | -------------- |
| upstream              | All           | semanticTokens | True    | 100     | 2.00    | 0.00  | 128.78 | 1.27     | 3.41      | 0.63        | 130.06 | 12         | 10           | 1364         | 5914       | 31271     | 102         | 296MB        | 568821MB       |
| cache-semantic-lookup | All           | semanticTokens | True    | 100     | 1.92    | 0.00  | 126.47 | 2.11     | 3.43      | 0.62        | 128.60 | 12         | 10           | 1364         | 5914       | 31271     | 102         | 290MB        | 562533MB       |
|                       |               |                |         |         |         |       |        |          |           |             |        |            |              |              |            |           |             |              |                |


But GetSemanticRules stably gives better result in a row
image

Another result.
image

**Comment by soulomoon** [comment]
After update to master 

| version               | configuration | name           | success | samples | startup | setup | userT  | delayedT | 1stBuildT | avgPerRespT | totalT | rulesBuilt | rulesChanged | rulesVisited | rulesTotal | ruleEdges | ghcRebuilds | maxResidency | allocatedBytes |
| --------------------- | ------------- | -------------- | ------- | ------- | ------- | ----- | ------ | -------- | --------- | ----------- | ------ | ---------- | ------------ | ------------ | ---------- | --------- | ----------- | ------------ | -------------- |
| upstream              | All           | semanticTokens | True    | 100     | 1.99    | 0.00  | 148.55 | 2.80     | 3.75      | 0.73        | 151.36 | 12         | 10           | 1364         | 5914       | 31271     | 102         | 307MB        | 594060MB       |
| cache-semantic-lookup | All           | semanticTokens | True    | 100     | 1.93    | 0.00  | 134.02 | 1.15     | 6.84      | 0.64        | 135.18 | 12         | 10           | 1364         | 5914       | 31271     | 102         | 318MB        | 562002MB       |
|                       |               |                |         |         |         |       |        |          |           |             |        |            |              |              |            |           |             |              |                |

**Comment by soulomoon** [comment]
Another run, the overal result is similar, but the detailed traces comparison indeed show some decent improvement.
| version  | configuration | name                           | success | samples | startup | setup | userT  | delayedT | 1stBuildT | avgPerRespT | totalT | rulesBuilt | rulesChanged | rulesVisited | rulesTotal | ruleEdges | ghcRebuilds | maxResidency | allocatedBytes |
| -------- | ------------- | ------------------------------ | ------- | ------- | ------- | ----- | ------ | -------- | --------- | ----------- | ------ | ---------- | ------------ | ------------ | ---------- | --------- | ----------- | ------------ | -------------- |
| upstream | All           | semanticTokens                 | True    | 50      | 1.83    | 0.00  | 64.00  | 1.18     | 3.50      | 0.62        | 65.18  | 12         | 10           | 1364         | 5914       | 31271     | 52          | 298MB        | 287351MB       |
| HEAD     | All           | semanticTokens                 | True    | 50      | 1.91    | 0.00  | 64.61  | 0.70     | 3.60      | 0.62        | 65.32  | 12         | 10           | 1364         | 5914       | 31271     | 52          | 299MB        | 285555MB       |


Screenshot 2024-02-18 at 12 50 39

image

image

**Comment by soulomoon** [comment]
cc @wz1000 , I have produce some more accurate result, seems about 5% improvement to `GetSemanticTokens` Rule.

File: shake-bench/src/Development/Benchmark/Rules.hs [hunk 1]
```diff
@@ -333,7 +333,7 @@ benchRules build MkBenchRules{..} = do
              ++ concat
                 [[ "-h"
                   , "-i" <> show i
-                  , "-po" <> outHp
+                  , "-po" <> dropExtension outHp
```
Comment (soulomoon) [inline shake-bench/src/Development/Benchmark/Rules.hs:336]: In macos, the original produce duplicated suffix `*.hp.hp`

**Comment by michaelpj** [comment]
> I have produced some more accurate result, seems about 5% improvement to GetSemanticTokens Rule.
Should be more if we consider the tokenizer alone.

FWIW I'm unsure if that's a big enough difference to make this worth it. It might be better just to keep the simpler code (we should definitely add the benchmark, though). Is the improvement bigger on bigger files?

**Comment by soulomoon** [comment]
> > I have produced some more accurate result, seems about 5% improvement to GetSemanticTokens Rule.
> 
> Should be more if we consider the tokenizer alone.
> 
> 
> 
> FWIW I'm unsure if that's a big enough difference to make this worth it. It might be better just to keep the simpler code (we should definitely add the benchmark, though). Is the improvement bigger on bigger files?

The bench is running on files with 2.5k and 1k lines of code.

agree，not much different since the current computation is relatively simple, perhaps I should turn it into draft and revisit this if/when more features are introduced that make the computation heavy.

**Pull request closed without merging**